doc/architecture/decisions/0015-key-management.md

```diff
@@ -0,0 +1,222 @@
+
+Secp256k1 wasn’t designed for encryption, but for signing, so at least some of our seeds and keys need to be generated using other standards.
+
+### Illustrative Workflows
+
+##### 1 - new / first holoport
+
+```rust
+// generate new random seed hierarchy
+let hd = DpkiHd::new_random();
+
+// display the primary/root seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_primary_seed()));
+
+// publish to the dpki dht
+let app_pub_data = DpkiApp::new_defaults(hd);
+publish_to_dpki_dht(app_pub_data);
+
+// initialize a device structure with device seed number 1
+let device = DpkiDevice::new_from_seed(hd.get_device_seed(1));
+// (optional) display device seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_device_seed(1)));
+// blink the device pin on the holoport circle
+display_to_user(device.get_device_pin());
+
+// reboot will require passphrase re-entry
+save_to_disk(device.get_saved(prompt_user_for_passphrase()));
+```
+
+##### 2 - existing holoport identity to a new box (device mnemonic)
+
+```rust
+// generate device dpki from user input
+let device = DpkiDevice::new_from_seed_and_pin(
+    Dpki::mnemonic_to_seed(prompt_user_for_device_seed()),
+    prompt_user_for_pin());
+
+// reboot will require passphrase re-entry
+save_to_disk(device.get_saved(prompt_user_for_passphrase()));
+```
+
+##### 3 - existing holoport identity to a new box (primary/root mnemonic)
+
+```rust
+// generate seed hierarchy from backup primary/root seed
+let hd = DpkiHd::new_from_primary_seed(
+    Dpki::mnemonic_to_seed(prompt_user_for_primary_seed()));
+
+// generate device dpki from user input
+let device = DpkiDevice::new_from_seed_and_pin(
+    hd.get_device_seed(prompt_user_for_device_index()),
+    prompt_user_for_pin());
+
+// reboot will require passphrase re-entry
+save_to_disk(device.get_saved(prompt_user_for_passphrase()));
+```
+
+##### 4 - new box / device generated keys - connect to existing dpki identity
+
+```rust
+// generate new random seed hierarchy
+let hd = DpkiHd::new_random();
+
+// display the primary/root seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_primary_seed()));
+
+// attach to existing identitiy
+let existing_app_pub_data = DpkiApp::fetch(
+    prompt_user_for_dpki_id());
+existing_app_pub_data.request_authorize_new_keys(hd);
+// user will need to sign the new keys into their chain from
+// another device that is already authorized
+
+// initialize a device structure with device seed number 1
+let device = DpkiDevice::new_from_seed(hd.get_device_seed(1));
+// (optional) display device seed to user for backup
+display_to_user(Dpki::seed_to_mnemonic(hd.get_device_seed(1)));
+// blink the device pin on the holoport circle
+display_to_user(device.get_device_pin());
+
+// reboot will require passphrase re-entry
+save_to_disk(device.get_saved(prompt_user_for_passphrase()));
+```
+
+##### 5 - new holo identity
```

Comment:
I'm not clear what a *holo identity* is.
Seems we are assuming a (centralized?) Holo account, managed by DpkiHolo and connected to our Dpki identity??

Reply:
I don't think there's anything **centralized** about it, but if you are creating a DPKI account that you want to use to bind some of your devices and apps together under a set of seeds so you can regenerate the keys, that is a Holo identity.  We have to be able to have web browsing users also generate Holo identities even when they don't run Holochain or DPKI locally, if they want to be able to WRITE to an app. They need to generate their authoring keys, and authorize hosts with network keys, to operate on their behalf.

What makes you think it's centralized though? Each person still manages their keys/seeds... the software provides for them to bridge across multiple apps or devices they use to access Holo apps.

Reply:
Yeah, so a "default" public identity would consist of:
- revocation public key
- 5 device signing keys that are all authorized to add additional device signing keys
which behind the scenes are managed with an HD hierarchy

A "holo" public identity would consist of:
- a single public key that is authorized to revoke and add additional device signing keys
which behind the scenes is generated from a combination of public salt and password hash

Reply:
From the point of view of code in the **Holochain** library I'm a little worried about having a structure with the name Holo.  How about if the structure were actually: 
`DpkiIdentity` (not `DpkiHolo`) and the `new` functions take a string based parameter that identifies an identity class that actually gets factored into the determinism, i.e. `new_random_from_passphrase("HOLO","some passphrase here")`  That way we aren't tightly binding Holochain to Holo's use of DPKI and allowing/encouraging other uses DPKI solutions too.

Reply:
The reason I asked about centralized ID is that it initialized only form a password.. Not really understanding what's going on and why this is different the other cases...

So you say that creates a key from a password? Instead of using seed words you take the password, salt it and treat that as seed to get a key pair from? And the reason is I guess to have an identity that you can recover just with that password?

Reply:
@zippy visibility that i asked a question re: "market research" that was never responded to ^^